akka-stream/src/main/scala/akka/stream/scaladsl/BidiFlow.scala

```diff
@@ -60,22 +59,42 @@ final class BidiFlow[-I1, +O1, -I2, +O2, +Mat](
    */
   def atopMat[OO1, II2, Mat2, M](bidi: Graph[BidiShape[O1, OO1, II2, I2], Mat2])(
       combine: (Mat, Mat2) => M): BidiFlow[I1, OO1, II2, O2, M] = {
-    val newBidi1Shape = shape.deepCopy()
-    val newBidi2Shape = bidi.shape.deepCopy()
 
-    // We MUST add the current module as an explicit submodule. The composite builder otherwise *grows* the
-    // existing module, which is not good if there are islands present (the new module will "join" the island).
-    val newTraversalBuilder =
-      TraversalBuilder
-        .empty()
-        .add(traversalBuilder, newBidi1Shape, Keep.right)
-        .add(bidi.traversalBuilder, newBidi2Shape, combine)
-        .wire(newBidi1Shape.out1, newBidi2Shape.in1)
-        .wire(newBidi2Shape.out2, newBidi1Shape.in2)
+    def actualAtop() = {
+      val newBidi1Shape = shape.deepCopy()
+      val newBidi2Shape = bidi.shape.deepCopy()
+
+      // We MUST add the current module as an explicit submodule. The composite builder otherwise *grows* the
+      // existing module, which is not good if there are islands present (the new module will "join" the island).
+      val newTraversalBuilder =
+        TraversalBuilder
+          .empty()
+          .add(traversalBuilder, newBidi1Shape, Keep.right)
+          .add(bidi.traversalBuilder, newBidi2Shape, combine)
+          .wire(newBidi1Shape.out1, newBidi2Shape.in1)
+          .wire(newBidi2Shape.out2, newBidi1Shape.in2)
+
+      new BidiFlow(
+        newTraversalBuilder,
+        BidiShape(newBidi1Shape.in1, newBidi2Shape.out1, newBidi2Shape.in2, newBidi1Shape.out2))
+    }
 
-    new BidiFlow(
-      newTraversalBuilder,
-      BidiShape(newBidi1Shape.in1, newBidi2Shape.out1, newBidi2Shape.in2, newBidi1Shape.out2))
+    if (this eq BidiFlow.identity) {
+      if (combine eq Keep.right) // optimization by returning bidi if possible since we know M == Mat2
+        BidiFlow.fromGraph(bidi).asInstanceOf[BidiFlow[I1, OO1, II2, O2, M]]
+      else if (combine eq Keep.left)
+        BidiFlow.fromGraph(bidi).mapMaterializedValue(_ => NotUsed).asInstanceOf[BidiFlow[I1, OO1, II2, O2, M]]
+      else actualAtop()
```

Comment:
We could also optimize the `combine == anything else` case by feeding the combiner a `NotUsed` statically without actually adding the stage.

Reply:
I tried that but in the end gave up because I anyway ended up doing most of actualAtop but perhaps I just made a bad turn somewhere. I should give it another shot.

Reply:
Fixed that now, plus added more complete test coverage that the matval stuff actually works.